src/dbnode/persist/fs/seek_manager.go

```diff
@@ -498,6 +504,68 @@ func (m *seekerManager) updateOpenLeaseHotSwapSeekers(
 	return wg, updateOpenLeaseResult, nil
 }
 
+func (m *seekerManager) RelinquishShard(shard uint32) error {
+	byTime := m.seekersByTime(shard)
+
+	// Indicate this shard is closed so it can't be borrowed and no seekers
+	// can be opened from now on.
+	byTime.Lock()
+	byTime.closed = true
+	byTime.Unlock()
+
+	for _, seekers := range byTime.seekers {
+		for _, s := range []seekersAndBloom{seekers.active, seekers.inactive} {
+			anySeekersAreBorrowed := false
+			for _, seeker := range s.seekers {
+				if seeker.isBorrowed {
+					anySeekersAreBorrowed = true
+					break
+				}
+			}
+
+			if anySeekersAreBorrowed {
+				// If any of the seekers are borrowed setup a waitgroup which will be used to
+				// signal when they've all been returned (the last seeker that is returned via
+				// the Return() API will call wg.Done()).
+				byTime.Lock()
```

Comment:
If we hold the lock for longer as described above then this will become an unlock then you'll wait for the `wg` then you'll lock again when you're done waiting

Reply:
We need to unlock for for waiting, but not before setting a new `wg` though right?